AX: The Node pointer member of the TextMarkerData structure is no longer needed.

AndresGonzalezApple opened the pull request:
#### 3341c5cc298816c00e808c6b971a6537104a02e1
<pre>
AX: The Node pointer member of the TextMarkerData structure is no longer needed.
<a href="https://bugs.webkit.org/show_bug.cgi?id=276131">https://bugs.webkit.org/show_bug.cgi?id=276131</a>
&lt;<a href="https://rdar.apple.com/problem/130985322">rdar://problem/130985322</a>&gt;

Reviewed by Tyler Wilcock.

This change removes the Node pointer from the TextMarkerData structure since it is no longer needed. The treeID and objectID uniquely identify the AX object, and the Node can be obtained from the ax object when needed. This removes a great deal of complexity from the code since there is no need to keep track of Nodes referenced from a TextMarker any longer. It also eliminates the need for the RawTextMarkerData in addition to the TextMarkerData structure.

* LayoutTests/accessibility/mac/textmarker-routines.html:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::remove):
(WebCore::AXObjectCache::visiblePositionForTextMarkerData):
(WebCore::AXObjectCache::characterOffsetForTextMarkerData):
(WebCore::AXObjectCache::textMarkerDataForCharacterOffset):
(WebCore::AXObjectCache::objectForTextMarkerData):
(WebCore::AXObjectCache::textMarkerDataForVisiblePosition):
(WebCore::AXObjectCache::prepareForDocumentDestruction):
(WebCore::AXObjectCache::accessibilityObjectForTextMarkerData): Deleted.
(WebCore::AXObjectCache::textMarkerDataForFirstPositionInTextControl): Deleted.
* Source/WebCore/accessibility/AXObjectCache.h:
(WebCore::AXObjectCache::nodeForID const):
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::TextMarkerData::TextMarkerData):
(WebCore::AXTextMarker::operator VisiblePosition const):
(WebCore::AXTextMarker::operator CharacterOffset const):
(WebCore::AXTextMarker::debugDescription const):
(WebCore::AXTextMarkerRange::AXTextMarkerRange):
(WebCore::AXTextMarkerRange::intersectionWith const):
(WebCore::AXTextMarker::setNodeIfNeeded const): Deleted.
(WebCore::RawTextMarkerData::toTextMarkerData const): Deleted.
* Source/WebCore/accessibility/AXTextMarker.h:
(WebCore::TextMarkerData::TextMarkerData):
(WebCore::TextMarkerData::axTreeID const):
(WebCore::TextMarkerData::axObjectID const):
(WebCore::AXTextMarker::AXTextMarker):
(WebCore::AXTextMarker::treeID const):
(WebCore::AXTextMarker::objectID const):
(WebCore::RawTextMarkerData::RawTextMarkerData): Deleted.
(): Deleted.
(WebCore::TextMarkerData::toRawTextMarkerData const): Deleted.
(WebCore::AXTextMarker::node const): Deleted.
* Source/WebCore/accessibility/cocoa/AXTextMarkerCocoa.mm:
(WebCore::AXTextMarker::AXTextMarker):
(WebCore::AXTextMarker::platformData const):
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityTextMarker initWithData:cache:]):
(-[WebAccessibilityTextMarker dataRepresentation]):
(-[WebAccessibilityTextMarker accessibilityObject]):
(-[WebAccessibilityTextMarker description]):
* Source/WebCore/accessibility/isolatedtree/mac/AXIsolatedObjectMac.mm:
(WebCore::AXIsolatedObject::attributedStringForTextMarkerRange const):
* Source/WebCore/accessibility/mac/AXObjectCacheMac.mm:
(WebCore::addFirstTextMarker):
(WebCore::getBytesFromAXTextMarker):
(WebCore::accessibilityObjectForTextMarker):
(WebCore::textMarkerForVisiblePosition):
(WebCore::textMarkerForCharacterOffset):
(WebCore::startOrEndTextMarkerForRange):

Canonical link: <a href="https://commits.webkit.org/280689@main">https://commits.webkit.org/280689@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/edbe9eb3e7e1a27ce15eeedccac73c5cd05daed9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57171 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36499 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9646 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60792 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7614 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/59299 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44123 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7804 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46297 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5360 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59201 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34263 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49365 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27157 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31045 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6681 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6619 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/53003 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6951 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62472 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1084 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7052 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53558 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1089 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49404 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53627 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12671 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/918 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32328 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33413 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34498 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33159 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AXTextMarker.h:
```diff
@@ -181,14 +149,11 @@ class AXTextMarker {
     String debugDescription() const;
 
     // Sets m_data.node when the marker was created with a PlatformTextMarkerData that lacks the node pointer because it was created off the main thread.
-    void setNodeIfNeeded() const;
```
Can we remove this comment now too?

AndresGonzalezApple replied:
Of course. Thansk!

twilco commented on Source/WebCore/accessibility/AXTextMarker.h:
```diff
@@ -63,24 +60,23 @@ struct RawTextMarkerData {
     // Members initialization alone is not enough to guaranty that all bytes in the struct memeory are initialized, and may cause random inequalities when doing byte-comparisons.
-    // For an example of such byte-comparison, see the TestRunner WTR::AccessibilityTextMarker::isEqual.
-    RawTextMarkerData()
+    // For an exampel of such byte-comparison, see the TestRunner WTR::AccessibilityTextMarker::isEqual.
```
Misspelling here: "exampel"

AndresGonzalezApple replied:
Corrected.

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -3363,16 +3361,20 @@ CharacterOffset AXObjectCache::characterOffsetFromVisiblePosition(const VisibleP
         return nullptr;
 
-    return getOrCreate(*domNode);
+    Node* node = object->node();
```
Do we want to hold a strong reference to this node via RefPtr to make sure it stays valid for `getOrCreate`?

AndresGonzalezApple replied:
In getOrCreate we are already doing:
Ref protectedNode { node };
So it would be refcount churn. The question moving forward is, do we always want to guaranty the survival of the parameter in the caller or the callee.

twilco commented on LayoutTests/accessibility/mac/textmarker-routines.html:
```diff
@@ -35,7 +35,8 @@
         shouldBe("item2.indexForTextMarker(secondTextMarker)", "5");
-        shouldBeTrue("item2.textMarkerForIndex(item2.indexForTextMarker(secondTextMarker)).isEqual(secondTextMarker)");       
+        shouldBeTrue("item2.textMarkerForIndex(item2.indexForTextMarker(secondTextMarker)).isEqual(secondTextMarker)");
+
```
Seems like we added an unnecessary new line here?

AndresGonzalezApple replied:
Yes, I removed the trailing spaces in this line of code and ended up adding an extra blank line while debugging test failures :-). Deleted a few unnecessary blank lines while at it. This test needs to be brought to AX standard but that's a matter for another patch.

AndresGonzalezApple committed:
AX: The Node pointer member of the TextMarkerData structure is no longer needed.
https://bugs.webkit.org/show_bug.cgi?id=276131
<rdar://problem/130985322>

Reviewed by Tyler Wilcock.

This change removes the Node pointer from the TextMarkerData structure since it is no longer needed. The treeID and objectID uniquely identify the AX object, and the Node can be obtained from the ax object when needed. This removes a great deal of complexity from the code since there is no need to keep track of Nodes referenced from a TextMarker any longer. It also eliminates the need for the RawTextMarkerData in addition to the TextMarkerData structure.

* LayoutTests/accessibility/mac/textmarker-routines.html:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::remove):
(WebCore::AXObjectCache::visiblePositionForTextMarkerData):
(WebCore::AXObjectCache::characterOffsetForTextMarkerData):
(WebCore::AXObjectCache::textMarkerDataForCharacterOffset):
(WebCore::AXObjectCache::objectForTextMarkerData):
(WebCore::AXObjectCache::textMarkerDataForVisiblePosition):
(WebCore::AXObjectCache::prepareForDocumentDestruction):
(WebCore::AXObjectCache::accessibilityObjectForTextMarkerData): Deleted.
(WebCore::AXObjectCache::textMarkerDataForFirstPositionInTextControl): Deleted.
* Source/WebCore/accessibility/AXObjectCache.h:
(WebCore::AXObjectCache::nodeForID const):
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::TextMarkerData::TextMarkerData):
(WebCore::AXTextMarker::operator VisiblePosition const):
(WebCore::AXTextMarker::operator CharacterOffset const):
(WebCore::AXTextMarker::debugDescription const):
(WebCore::AXTextMarkerRange::AXTextMarkerRange):
(WebCore::AXTextMarkerRange::intersectionWith const):
(WebCore::AXTextMarker::setNodeIfNeeded const): Deleted.
(WebCore::RawTextMarkerData::toTextMarkerData const): Deleted.
* Source/WebCore/accessibility/AXTextMarker.h:
(WebCore::TextMarkerData::TextMarkerData):
(WebCore::TextMarkerData::axTreeID const):
(WebCore::TextMarkerData::axObjectID const):
(WebCore::AXTextMarker::AXTextMarker):
(WebCore::AXTextMarker::treeID const):
(WebCore::AXTextMarker::objectID const):
(WebCore::RawTextMarkerData::RawTextMarkerData): Deleted.
(): Deleted.
(WebCore::TextMarkerData::toRawTextMarkerData const): Deleted.
(WebCore::AXTextMarker::node const): Deleted.
* Source/WebCore/accessibility/cocoa/AXTextMarkerCocoa.mm:
(WebCore::AXTextMarker::AXTextMarker):
(WebCore::AXTextMarker::platformData const):
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityTextMarker initWithData:cache:]):
(-[WebAccessibilityTextMarker dataRepresentation]):
(-[WebAccessibilityTextMarker accessibilityObject]):
(-[WebAccessibilityTextMarker description]):
* Source/WebCore/accessibility/isolatedtree/mac/AXIsolatedObjectMac.mm:
(WebCore::AXIsolatedObject::attributedStringForTextMarkerRange const):
* Source/WebCore/accessibility/mac/AXObjectCacheMac.mm:
(WebCore::addFirstTextMarker):
(WebCore::getBytesFromAXTextMarker):
(WebCore::accessibilityObjectForTextMarker):
(WebCore::textMarkerForVisiblePosition):
(WebCore::textMarkerForCharacterOffset):
(WebCore::startOrEndTextMarkerForRange):

Canonical link: https://commits.webkit.org/280689@main
